File: framework/backend/python/packages/k1s0-cache/src/k1s0_cache/patterns/write_through.py
"""Write-through caching pattern."""

from __future__ import annotations

from collections.abc import Awaitable, Callable

from k1s0_cache.operations import CacheOperations


class WriteThrough:
    """Writes to the backing store first, then updates the cache.

    Args:
        cache: Cache operations instance.
        default_ttl: Default TTL applied when none is specified per call.
    """

    def __init__(
        self, cache: CacheOperations, default_ttl: int | None = None
    ) -> None:
        self._cache = cache
        self._default_ttl = default_ttl

    async def write(
        self,
        key: str,
        value: str,
        writer: Callable[[str, str], Awaitable[None]],
        ttl: int | None = None,
    ) -> None:
        """Write to the backing store and then update the cache.

        Args:
            key: The cache key.
            value: The value to write.
            writer: An async callable that persists the key-value pair.
            ttl: Optional TTL override.
        """
        await writer(key, value)
        effective_ttl = ttl if ttl is not None else self._default_ttl
        await self._cache.set(key, value, ttl=effective_ttl)

File: framework/backend/csharp/src/K1s0.Cache/Patterns/WriteThrough.cs
namespace K1s0.Cache.Patterns;

/// <summary>
/// Implements the write-through pattern: writes go to the backing store first, then to the cache.
/// </summary>
public sealed class WriteThrough
{
    private readonly ICacheOperations _cache;
    private readonly TimeSpan? _defaultTtl;

    /// <summary>
    /// Initializes a new instance of the <see cref="WriteThrough"/> class.
    /// </summary>
    /// <param name="cache">The cache operations instance.</param>
    /// <param name="defaultTtl">Optional default TTL for cached entries.</param>
    public WriteThrough(ICacheOperations cache, TimeSpan? defaultTtl = null)
    {
        _cache = cache ?? throw new ArgumentNullException(nameof(cache));
        _defaultTtl = defaultTtl;
    }

    /// <summary>
    /// Writes the value to the backing store via the writer delegate, then stores it in the cache.
    /// </summary>
    /// <param name="key">The cache key.</param>
    /// <param name="value">The value to write.</param>
    /// <param name="writer">A delegate that persists the key-value pair to the backing store.</param>
    /// <param name="ttl">Optional TTL override.</param>
    /// <param name="ct">A cancellation token.</param>
    public async Task WriteAsync(
        string key,
        string value,
        Func<string, string, Task> writer,
        TimeSpan? ttl = null,
        CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(writer);

        // Write to backing store first.
        await writer(key, value).ConfigureAwait(false);

        // Then update the cache.
        await _cache.SetAsync(key, value, ttl ?? _defaultTtl, ct).ConfigureAwait(false);
    }
}
